### DISCUSSION TITLE
[vcpkg|manifest] Missing a way to pass --head to certain dependencies.

### DISCUSSION THREAD

**Discussion by Neumann-A** [microsoft/vcpkg#17805] [Ideas]
The vcpkg.json should probably have an extra field `vcpkg_additional_args` or similar to pass additional vcpkg arguments to the build of some dependencies

**Comment by strega-nil** [comment]
Is there any other reason you want this besides passing `--head`?

**Comment by Neumann-A** [comment]
@strega-nil: currently cannot think of anything more. Just always want to install `eigen3 --head` and probably a few other dependencies. Especially if the ports in questions are your own ports via a overlay you might want to use `--head` for development without always changing the ref&hash

**Comment by strega-nil** [comment]
Would it be okay to add something like a `"x-extra-arguments": [ "head" ]` to the dependency field?

**Comment by strega-nil** [comment]
Or maybe in `vcpkg-configuration.json`, `x-dependency-arguments: { "eigen3": [ "head" ] }`

**Comment by sylveon** [comment]
I would also want this, to always install the head of `wil` and `rapidjson`, however I feel that `vcpkg-configuration` is not a place where I'd want to put it in manifest mode, since it directly concerns my project, not vcpkg's configuration

**Comment by strega-nil** [comment]
@sylveon `vcpkg-configuration.json` is the project configuration that doesn't belong in `vcpkg.json`: in other words, the stuff that people depending on you shouldn't care about.

**Comment by sylveon** [comment]
My project won't build with the very old stable version of RapidJSON, so knowing that it's head should matter, no?

**Comment by strega-nil** [comment]
it's one of those things that versioning should fix; using `head` is not supported and often doesn't work, and so it's not a part of the dependency interface of your library.

**Comment by sylveon** [comment]
I've skimmed through the versioning docs and can't seem to find anything related to installing the head of a library. Is that feature not present at the moment?

**Comment by strega-nil** [comment]
poke @vicroms ^

**Comment by sylveon** [comment]
Alternatively, instead of `head`, is it possible to pin a dep to an arbitrary commit with versioning?

**Comment by RomanValov** [comment]
+1 for a way to pass head via manifest. My use case is private lib and private registry.
Private lib is actively changed during the application development.
Thus I would like to commit to my private lib and re-configure and re-build my dependent app.
Instead every time I have to copy paste commit hash id into registry portfile.

**Comment by spinicist** [comment]
Was this functionality ever implemented? I have a similar use case to @RomanValov

**Comment by gix** [comment]
This is not only useful for private ports. For example Detours had its last release somewhere in 2016 (used by its port), but has seen considerable maintenance since then to fix issues with newer compilers.

**Reply by Neumann-A** [reply]
> For example Detours had its last release somewhere in 2016 (used by its port), but has seen considerable maintenance since then to fix issues with newer compilers.

Then update the port and use version-date. Pointing vcpkg to releases only is not mandatory. It is preferred but not a hard requirement.

**Reply by sylveon** [reply]
This is my exact use case actually

**Comment by jpgaribotti** [comment]
I have the same use case as @RomanValov and @spinicist: developping an application and library in a private registry in parallel. Which seems a pretty normal situation, particularly in professional environments. I want to move to vcpkg, but without support for this workflow it would be difficult. I'm using scripts to facilitate it, but it's still less than ideal.

**Comment by QuercusFelis** [comment]
This would be a very convenient feature for magnum, which runs on a rolling release, but whose port files currently reference a release tag that's 5 years old now. They only support installation through classic mode using --head. Dropping a link to the relevant magnum issue here for reference.
https://github.com/mosra/magnum-integration/issues/118

**Reply by sylveon** [reply]
There are a few other rolling release ports in vcpkg, and people tend to just update it when they need a new thing. Take a look at the wil port.

**Reply by QuercusFelis** [reply]
@sylveon I don't really want to be responsible for updating and maintaining the port as a user. I am just trying to bring magnum into my project and have been able to use vcpkg manifests without issue for everything else that has a port file so far. The official workaround for magnum is to just do things the old fashioned way with a cmake subproject, or using CPM, which kind of defeats the whole point of using vcpkg for me when a simple argument I could pass in the manifest file would unblock me the fastest (if such a feature existed).

**Reply by sylveon** [reply]
You aren't expected to become an owner of the port. Most ports are community contributed. I am suggesting a one-off PR that uses a git SHA as the object to checkout instead of a tag, and sets the version to the current date.

**Reply by spinicist** [reply]
It's not too much effort to make an overlay port so you can keep Magnum at a commit you are happy with. This is what I do with Eigen.

However, this is very much a deficiency as there is something you can do in classic mode that you can't do in manifest mode.

**Reply by Osyotr** [reply]
`--head` is not a real solution to rolling release problem. It won't save you from changes in upstream buildsystem, including changes in dependency tree. So you'll end up making an overlay port sooner or later.

**Reply by spinicist** [reply]
Then why is it included in classic mode?

I get it, there is always going to be a conflict between versioned ports and rolling releases. What would be amazing, in my opinion, is the ability to specify a commit ID in the `overrides` section of `vcpkg.json`. Overlay ports are an extra level of effort.
